modules/terraform/test_helpers_test.go

```diff
@@ -0,0 +1,66 @@
+package terraform
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestUnitTestSimple(t *testing.T) {
+	testConfig := &UnitTestConfig{
+		TerraformOptions: &Options{
+			TerraformDir: "../../examples/terraform-hello-world-example/",
+		},
+		Validators: []Validator{
+			validateOutput,
+		},
+	}
+
+	UnitTest(t, testConfig)
+}
```

Comment:
Sanity check: how does it compare with the "hello, world" example we have today (e.g., [on this page](https://terratest.gruntwork.io/examples/))? E.g.,: 

- Is it shorter? 
- Easier to understand for newbies? 
- Less error prone?

Reply:
It's definitely not shorter, nor easier to understand for a newbie to terratest, it also wouldn't teach them the very basics of what's going on under the hood.

I might argue it's less error prone since they can't forget to destroy the infrastructure they setup, and they are freely provided the outputs without having to call for them.

However although it's slightly longer and harder than the very basic hello world example the strength comes when you have more tests, more complex modules, and especially when compared with the simplest examples of using test stages (especially since it will deal with saving the outputs and reloading them if you are skipping the init_apply step.

I used the hello world example here since it was the simplest way to flex the few things I wanted to do, although the example probably shouldn't be in the quick start guide, since it doesn't teach the newbie what's going on.

I _think_ (but I haven't yet written much in the way of integration tests) that the integration test version should be especially powerful, and much simpler than orchestrating all of the interactions yourself (especially when combined with the test stages).

Reply:
I think also someone reading later reasoning about the tests will find it easier to understand what tests are running if the boilerplate includes a nice long list of tests, like

```
TerraformOptions: &Options{
  TerraformDir: "../../examples/terraform-hello-world-example/",
},
Validators: []Validator{
  validateOutput,
  validateRds,
  validateAlb,
  validateRedis,
  validateEcsTask
  validateEcsService,
  validateCloudMap
},
```

It also sets a patter for function signatures which might encourage reuse of validators (although that is 100% conjecture on my part)

Reply:
> I used the hello world example here since it was the simplest way to flex the few things I wanted to do, although the example probably shouldn't be in the quick start guide, since it doesn't teach the newbie what's going on.

There are a large number of unit and integration test examples in the `test` folder... And in other open source `gruntwork-io` repos (e.g., https://github.com/hashicorp/terraform-aws-vault, https://github.com/hashicorp/terraform-aws-consul)... Do you want to try out the pattern on something more real-world to see if it scales better to more complicated use cases?

Reply:
(And by "try out the pattern", I mean, write up a few more examples just to see what they would look like with the new API)

Reply:
Sure thing

Reply:
A real problem created here is scope.....

....with the current scheme it's easy to set up expected values, use them to initialise the vars of the module, and then reuse them later.

If we break out the validations into external functions that scoping is lost and we need to communicate those things around (test_structure.Load/Save* can help, but it's just adding more boiler plate), since it's going to automatically save the terraformOptions you could load them again with test_structure.Load/Save (which save you doing the save yourself, but still adds a lot of boiler plate of loading everything you expect in every test).

You can define the validators as anonymous functions inside the TestConfig struct and keep those expectations in scope, but doing so makes a long struct, which might be hard to read, ends up with multiple levels of extra indentation and definitely breaks the advantage I mentioned earlier with being able to easily see a nice list of validators.

There could be an extra item in the Config (something like `TestData map[string]string` which could be passed around for you, but now the function signatures are getting quite long needed 4 args for every validator, and in the integration test this would become _another_ map of module names to TestData.

Reply:
Ah, that's a good point. And without generics, you also lose type safety with those maps, so you'd have to cast everything. 

Perhaps this is a terrible idea, but I wonder if you could declare a `struct` that contains all the values you care about, and that single struct is passed to each function instead of a `map`?

Reply:
Do you mean to allow the user to define a struct to pass around, like (in their client code, not terratest code):

```
type MyStruct struct {
   Region string
   Foo string
   Bar []int
}
```

Then pass it back to them on every test so they only have to:

```
func validateFoo(t *testing.T, outputs ModuleOutputs, testData interface{}, workingDir string) {
  myStruct := testData.(*MyStruct)
  assert.Equal(t, myStruct.Foo, outputs["foo"])
}
```

Even doing this you lost the compile time type safety, the type error only surfaces when the execution occurs :S

Reply:
Yes, and, yes, we lose type safety... But at least we reduce it to one cast per method. But yea, still not great.

Reply:
I've added a dynamodb test (which is a lot of copy+paste from https://github.com/gruntwork-io/terratest/blob/master/test/terraform_aws_dynamodb_example_test.go) and refactored to use the new interface (excluding passing the struct around we just discussed, with a Save/LoadString workaround.

I'll update the docs now to reflect the struct discussion we just had.

Reply:
I've also updated the docs (for UnitTest, not Integration test yet)

Reply:
Updated the Dynamo test now to include the new test data struct